src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -123,12 +122,25 @@ protected override Payload ConstructHeadersPayload(GetHeadersPayload getHeadersP
                 return null;
             }
 
-            var headers = new ProvenHeadersPayload();
-            foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+            var provenHeadersPayload = new ProvenHeadersPayload();
+
+            // Do not return more than 2000 headers from the fork point.
+            ChainedHeader chainedHeaderToStartFrom = this.consensusManager.Tip;
+            if ((this.consensusManager.Tip.Height - fork.Height) > MaxItemsPerHeadersMessage)
```

Comment:
Why take an instance of chainedHeaderToStartFrom and not use it here in the `if`?

Reply:
I think there is a -1 problem. 

If CM.Tip is 5k, fork is 3k we will have

`if (5000 - 3000 > 2000)`

so the if won't be executed. 
Then we will collect 1999 headers instead of 2000

Reply:
Oh yes  of course,  fixed 👍